File: frontend/src/app/core/utils/slots.ts
import type { Booking, WorkingHours } from '../models/api.models';
import { isoToWeekday, minutesToTime, timeToMinutes } from './datetime';

/** A candidate start time, with the end time the backend will derive. */
export interface TimeSlot {
  /** "HH:MM", suitable for display and for `+ ':00'` to reach the API format. */
  start: string;
  /** "HH:MM" — what the server will store as `end_time`. */
  end: string;
  /** False when the slot collides with an existing non-cancelled booking. */
  available: boolean;
  /** Why the slot is unavailable, so the UI can say more than "no". */
  reason: 'booked' | null;
}

/** Slots are offered on this grid rather than at every minute. */
export const DEFAULT_SLOT_STEP_MINUTES = 15;

export interface SlotOptions {
  /** The staff member's hours for the chosen weekday, if any. */
  workingHours: WorkingHours | null | undefined;
  /** Service duration in minutes — the backend uses this to derive end_time. */
  durationMinutes: number;
  /** "YYYY-MM-DD". */
  date: string;
  /** Every booking already loaded for the organization. */
  bookings: Booking[];
  /** Staff member being booked. */
  staffId: number;
  /** Grid spacing. Defaults to 15 minutes. */
  stepMinutes?: number;
  /** Minutes after midnight below which slots are dropped (e.g. "now"). */
  notBeforeMinutes?: number | null;
}

/**
 * Compute the bookable time slots for one staff member on one date.
 *
 * This deliberately reproduces the backend's own arithmetic from
 * `bookings/serializers.py` so the picker and the server agree:
 *
 *   working hours : valid  <=>  start >= wh.start_time  AND  end <= wh.end_time
 *                   (otherwise "Booking time is outside the staff member's
 *                    working hours.")
 *
 *   overlap       : conflict <=> start < booking.end_time
 *                                AND end > booking.start_time
 *                   over bookings for the same staff and date, excluding those
 *                   with status CANCELLED
 *                   (otherwise "Staff member already has a booking during
 *                    this time.")
 *
 * Two consequences worth stating:
 *  - `end_time` is never an input. It is `start + duration`, exactly as the
 *    server computes it, and is shown only as a preview.
 *  - This is a convenience, not an authority. The server re-validates on
 *    submit and may still reject a slot — for instance if someone else booked
 *    it in the meantime. Callers must surface that rejection.
 *
 * Returns an empty list when there is nothing sensible to offer (no hours
 * configured for that weekday, no duration, or an unparseable date).
 */
export function computeTimeSlots(options: SlotOptions): TimeSlot[] {
  const {
    workingHours,
    durationMinutes,
    date,
    bookings,
    staffId,
    stepMinutes = DEFAULT_SLOT_STEP_MINUTES,
    notBeforeMinutes = null,
  } = options;

  if (!workingHours || !workingHours.is_available) {
    return [];
  }
  if (!durationMinutes || durationMinutes <= 0) {
    return [];
  }
  if (isoToWeekday(date) === null) {
    return [];
  }

  const dayStart = timeToMinutes(workingHours.start_time);
  const dayEnd = timeToMinutes(workingHours.end_time);
  if (dayStart === null || dayEnd === null) {
    return [];
  }

  // Busy intervals for this staff member on this date. CANCELLED bookings free
  // the slot, matching the backend's `.exclude(status=CANCELLED)`.
  const busy: Array<[number, number]> = bookings
    .filter(
      (booking) =>
        booking.staff === staffId &&
        booking.booking_date === date &&
        booking.status !== 'CANCELLED',
    )
    .map((booking): [number, number] => [
      timeToMinutes(booking.start_time) ?? 0,
      timeToMinutes(booking.end_time) ?? 0,
    ])
    .filter(([start, end]) => end > start);

  const earliest =
    notBeforeMinutes !== null && notBeforeMinutes > dayStart ? notBeforeMinutes : dayStart;

  const slots: TimeSlot[] = [];
  // `start + duration` must fit inside the working day, so the last candidate
  // is `dayEnd - duration`.
  for (let start = earliest; start + durationMinutes <= dayEnd; start += stepMinutes) {
    const end = start + durationMinutes;
    const booked = busy.some(([busyStart, busyEnd]) => start < busyEnd && end > busyStart);

    slots.push({
      start: minutesToTime(start),
      end: minutesToTime(end),
      available: !booked,
      reason: booked ? 'booked' : null,
    });
  }

  return slots;
}

/** Convenience for templates: how many slots are actually bookable. */
export function countAvailable(slots: readonly TimeSlot[]): number {
  return slots.reduce((total, slot) => (slot.available ? total + 1 : total), 0);
}

File: backend/availability/slots.py
"""Availability slot computation.

Mirrors `frontend/src/app/core/utils/slots.ts`, which itself mirrors the
arithmetic in `bookings/serializers.py`, so the picker, this endpoint and the
booking validation all agree:

    working hours : valid  <=>  start >= wh.start_time
                                AND start + duration <= wh.end_time

    overlap       : conflict <=> start < booking.end_time
                                 AND end > booking.start_time
                    over bookings for the same staff and date, excluding those
                    with status CANCELLED

The server remains the authority: this helper only decides what to *offer*.
`BookingSerializer.validate()` re-checks everything on submit.
"""

from datetime import datetime, time

SLOT_STEP_MINUTES = 15


def time_to_minutes(value):
    """`datetime.time` -> minutes after midnight, or None."""
    if value is None:
        return None
    return value.hour * 60 + value.minute


def minutes_to_time(minutes):
    """Minutes after midnight -> `datetime.time`."""
    return time(hour=minutes // 60, minute=minutes % 60)


def compute_slots(
    working_hours,
    duration_minutes,
    booking_date,
    busy_bookings,
    now=None,
    step_minutes=SLOT_STEP_MINUTES,
):
    """
    Return the slots offered for one staff member on one date.

    `working_hours`      the WorkingHours row for that staff and weekday, or None
    `duration_minutes`   the service duration; also the implied end offset
    `booking_date`       a `datetime.date`
    `busy_bookings`      bookings for that staff and date, already excluding
                         CANCELLED (the caller decides, so the rule stays in one
                         place next to the booking query)
    `now`                optional aware-or-naive datetime; when supplied, slots
                         that have already started today are not offered

    Each slot is `{"start": time, "end": time, "available": bool, "reason": ...}`.
    """

    if working_hours is None or not working_hours.is_available:
        return []

    if not duration_minutes or duration_minutes <= 0:
        return []

    day_start = time_to_minutes(working_hours.start_time)
    day_end = time_to_minutes(working_hours.end_time)

    if day_start is None or day_end is None:
        return []

    busy = []
    for booking in busy_bookings:
        start = time_to_minutes(booking.start_time)
        end = time_to_minutes(booking.end_time)
        if start is None or end is None or end <= start:
            continue
        busy.append((start, end))

    earliest = day_start

    # Do not offer slots that have already begun today. This only ever removes
    # options; the past-date rule in BookingSerializer still applies on submit.
    if now is not None and now.date() == booking_date:
        now_minutes = now.hour * 60 + now.minute
        if now_minutes > earliest:
            # Round up to the grid so the offered times stay on 15-minute marks.
            earliest = -(-now_minutes // step_minutes) * step_minutes

    slots = []
    start = earliest

    while start + duration_minutes <= day_end:
        end = start + duration_minutes
        booked = any(start < busy_end and end > busy_start for busy_start, busy_end in busy)

        slots.append(
            {
                "start": minutes_to_time(start),
                "end": minutes_to_time(end),
                "available": not booked,
                "reason": "booked" if booked else None,
            }
        )

        start += step_minutes

    return slots
